### ISSUE TITLE
To Sign Existing Private Key using XADESJS

### ISSUE THREAD

**Issue by erdembas** [PeculiarVentures/xadesjs#36]
Hello,

How can I sign xml document with existing privatekey using xadesjs?

My smart card doesn't support KeyPairGen!

My smart card's mechanisms;
![image](https://cloud.githubusercontent.com/assets/2093046/17935813/4ced5102-6a25-11e6-91d3-5013f02afa47.png)

Thanks

**Comment by rmhrisk** [comment]
@erdembas glad to see your looking at this library aslo :) To use your smart card and xadesjs you will need https://github.com/PeculiarVentures/node-webcrypto-p11/, it is a layer on graphene that exposes the capabilities of the HSM via the WebCrypto interfaces.

You have two problems when using an existing key:
1. Discovering a pre-existing key,
2. Getting the certificate associated with the key.

Now the WebCrypto interfaces do not have either of these concepts. We implemented an extension to the interfaces derived from a proposed draft called [KeyStorage](https://github.com/PeculiarVentures/node-webcrypto-p11/blob/master/lib/key_storage.ts#L31). As implemented right now it lets you generate a key and find it later. It does this by associating a label with the key that contains data about it, this data is in JSON form.

This is a bit of a problem for your specific use case as you want to use a pre-existing key that was not generated with node-webcrypto-p11. This is of course, a legit use case, we just had not considered it because we didn't need it. Adding @microshine as we need to discuss how to support this hopefully in a consistent way so existing keys generated by us and those generated by others can be discoverable.

The second issue is certificate discovery, we also presumed in our usage we would store the certificates outside of the security device, however in a smart card usage case like yours I think you would want to be able to get the certificate out via this same interface.

In short, right now this isn't supported by node-webcrypto-p11, we are willing to take the time to make it work if you will work with us as we do it.

One possible way to find the key would be by adding:

```
getItem(key: string, alg?: Algorithm): CryptoKey
```

This would allow you to search for keys based on labels, would that work for your use case? Do your keys have unique labels?

Also this bug really should be in node-webcrypto-p11, mind closing this one and opening one over there linking to this conversation?

**Comment by erdembas** [comment]
@rmhrisk I will be glad for me to contribute to this library :) 
The code belowe can be used for this library.

I got existing PrivateKey as converting CryptoKey successfully. Added key_storage.js method getCryptoKeyItemByClass, getCryptoKeyItemByLabel etc. 

   ```
 KeyStorage.prototype.getItemByClass = function (id) {
        var keys = this.session.find({ class: id });
        if (!keys.length) {
            return null;
        }
        if (keys.length > 1)
            console.log("WebCrypto:PKCS11: " + keys.length + " keys matches ID '" + id + "'");
        return keys.items(0);
    };

    KeyStorage.prototype.getCryptoKeyItemByClass = function (key, alg) {
        var sobj = this.getItemByClass(key);
        if (sobj) {
            var _key = sobj.toType();
            return new key_1.CryptoKey(_key, alg);
        }
        else
            return null;
    };
```

```
//Example code
var crypto = xadesjs.Application.crypto;

var mod = crypto.module;
var session = crypto.session;

var publicKey;
var privateKey;

privateKey = crypto.keyStorage.getCryptoKeyItemByClass(graphene.ObjectClass.PRIVATE_KEY, { name: "RSA_PKCS", hash: { name: "SHA-256" } });
publicKey = crypto.keyStorage.getCryptoKeyItemByClass(graphene.ObjectClass.PUBLIC_KEY, { name: "RSA_PKCS", hash: { name: "SHA-256" } });
```
CryptoKey's;
![image](https://cloud.githubusercontent.com/assets/2093046/17973599/bd198e2a-6aeb-11e6-942d-fe1fa63d1599.png)


xadesjs library uses default mechanism SHA_256_RSA_PKCS, but my smart card's support mechanism is RSA_PKCS. 

I have changed the code for using RSA_PKCS mechanism, but got sign exception as CKR_DATA_LEN_RANGE:33.

Exception Code Block;
![image](https://cloud.githubusercontent.com/assets/2093046/17973931/d9968c96-6aec-11e6-99e7-bc41aa6c7efd.png)

Exception;
![image](https://cloud.githubusercontent.com/assets/2093046/17973346/bb0d195e-6aea-11e6-928d-99f10adea54e.png)

Please helping me.

**Comment by microshine** [comment]
Hello. Could you check incomming data to `C_Sign` function?

> #define CKR_DATA_LEN_RANGE
> The plaintext input data to a cryptographic operation has a bad length. Depending on the operation's mechanism, this could mean that the plaintext data is too short, too long, or is not a multiple of some particular blocksize. This return value has higher priority than CKR_DATA_INVALID.

**Comment by erdembas** [comment]
How to get x509Certificate from PublicKey or CryptoKey?

**Comment by rmhrisk** [comment]
Certificates are stored as objects in PKCS11. You can enumerate the objects stored on your token with the graphene client.

Right now the node-webcrypto-p11 implementation does not expose the certificates.

You will need to use graphene directly to find it.

Ryan Hurst

**Comment by InvoiceFox** [comment]
Thank you greatly for writing this!

I am not sure I understand one thing. Could I be able to use Xadesjs with an existing certificate. For example one in pkcs12 format (parsing it with something like forge)?

**Comment by rmhrisk** [comment]
Our pleasure.

We have another library we create called PKIjs, you can think of it as a more standards compliant Forge, though in our opinions it has many more benefits also.

It does support PKCS#12; see this : https://unmitigatedrisk.com/?p=543 for the caveats.

Though we have not yet done the work to provide samples (and potentially make it easier to do this) you can access existing keys.

I would personally start with our https://github.com/PeculiarVentures/node-webcrypto-p11 project. This allows you, via a WebCrypto styled interface to access keys stored on a HSM.

You will need to discover the keys/certificates, for that you will need to use https://github.com/PeculiarVentures/graphene. Certificates are stored in PKCS#11 as "Objects" you can read those keys out.

**Comment by microshine** [comment]
@erdembas node-webcrypto-p11 supports `RSA_PKCS` now. It's not published yet. You can try it form branch `RSA_PKCS` https://github.com/PeculiarVentures/node-webcrypto-p11/pull/29
